File: languages/csharp/AppEncryption/AppEncryption.Tests/Crypto/Keys/SecretCryptoKeyTest.cs
using System;
using GoDaddy.Asherah.Crypto.Keys;
using GoDaddy.Asherah.SecureMemory;
using Moq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.Crypto.Keys
{
    [Collection("Logger Fixture collection")]
    public class SecretCryptoKeyTest
    {
        private const bool Revoked = false;

        private readonly DateTimeOffset creationTime = DateTimeOffset.UtcNow;

        private readonly Mock<Secret> secretMock;
        private readonly SecretCryptoKey secretCryptoKey;

        public SecretCryptoKeyTest()
        {
            secretMock = new Mock<Secret> { CallBase = true };
            secretCryptoKey = new SecretCryptoKey(secretMock.Object, creationTime, Revoked);
        }

        [Fact]
        public void TestConstructor()
        {
            Assert.NotNull(secretCryptoKey);
        }

        [Fact]
        public void TestConstructorUsingOtherKey()
        {
            Mock<Secret> secretToCopy = new Mock<Secret> { CallBase = true };
            Mock<SecretCryptoKey> secretCryptoKeyMock = new Mock<SecretCryptoKey>(null, creationTime, Revoked);
            secretCryptoKeyMock.SetupGet(x => x.Secret).Returns(secretToCopy.Object);
            secretCryptoKeyMock.Setup(x => x.GetCreated()).Returns(creationTime);
            secretToCopy.Setup(x => x.CopySecret()).Returns(secretMock.Object);

            SecretCryptoKey secretCryptoKey = new SecretCryptoKey(secretCryptoKeyMock.Object);

            Assert.NotNull(secretCryptoKey);
            Assert.Equal(creationTime, secretCryptoKey.GetCreated());
            Assert.Equal(secretMock.Object, secretCryptoKey.Secret);
            Assert.False(secretCryptoKey.IsRevoked());
        }

        [Fact]
        public void TestKeyWithFunction()
        {
            Func<byte[], object> action = keyBytes => null;
            secretCryptoKey.WithKey(action);
            secretMock.Verify(x => x.WithSecretBytes(action), Times.Once);
        }

        [Fact]
        public void TestKeyWithAction()
        {
            Action<byte[]> action = keyBytes => { };
            secretCryptoKey.WithKey(action);
            secretMock.Verify(x => x.WithSecretBytes(action), Times.Once);
        }

        [Fact]
        public void TestGetCreated()
        {
            DateTimeOffset actualCreated = secretCryptoKey.GetCreated();
            Assert.Equal(creationTime, actualCreated);
        }

        [Fact]
        public void TestMarkRevoked()
        {
            Assert.False(secretCryptoKey.IsRevoked());
            secretCryptoKey.MarkRevoked();
            Assert.True(secretCryptoKey.IsRevoked());
        }

        [Fact]
        public void TestClose()
        {
            secretCryptoKey.Dispose();
            secretMock.Verify(x => x.Dispose(), Times.Once);
        }

        [Fact]
        public void TestGetSecret()
        {
            Secret actualSecret = secretCryptoKey.Secret;
            Assert.Equal(secretMock.Object, actualSecret);
        }
    }
}

File: java/app-encryption/src/test/java/com/godaddy/asherah/crypto/keys/SecretCryptoKeyTest.java
package com.godaddy.asherah.crypto.keys;


import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.godaddy.asherah.securememory.Secret;

import java.time.Instant;
import java.util.function.Consumer;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class SecretCryptoKeyTest {

  @Mock
  private Secret secretMock;

  private SecretCryptoKey secretCryptoKey;
  private final Instant creationTime = Instant.now();
  private final boolean revoked = false;

  @BeforeEach
  void setUp() {
    secretCryptoKey = new SecretCryptoKey(secretMock, creationTime, revoked);
  }

  @Test
  void testConstructor() {
    assertNotNull(secretCryptoKey);
  }

  @Test
  void testConstructorUsingOtherKey() {
    Secret secretToCopy = mock(Secret.class);
    SecretCryptoKey secretCryptoKeyMock = mock(SecretCryptoKey.class);
    when(secretCryptoKeyMock.getSecret()).thenReturn(secretToCopy);
    when(secretToCopy.copySecret()).thenReturn(secretMock);
    when(secretCryptoKeyMock.getCreated()).thenReturn(creationTime);

    SecretCryptoKey secretCryptoKey = new SecretCryptoKey(secretCryptoKeyMock);

    assertNotNull(secretCryptoKey);

    Instant expectedCreationTime = creationTime;
    Instant actualCreationTime = secretCryptoKey.getCreated();
    assertEquals(expectedCreationTime, actualCreationTime);

    Secret actualSecret = secretCryptoKey.getSecret();
    assertEquals(secretMock, actualSecret);
  }

  @Test
  void testWithKeyFunction() {
    Function<byte[], ?> action = keyBytes -> null;
    secretCryptoKey.withKey(action);
    verify(secretMock).withSecretBytes(action);
  }

  @Test
  void testWithKeyConsumer() {
    Consumer<byte[]> action = bytes -> {
    };
    secretCryptoKey.withKey(action);
    verify(secretMock).withSecretBytes(action);
  }

  @Test
  void testGetCreated() {
    final Instant actualCreated = secretCryptoKey.getCreated();
    assertEquals(creationTime, actualCreated);
  }

  @Test
  void testMarkRevoked() {
    assertFalse(secretCryptoKey.isRevoked());
    secretCryptoKey.markRevoked();
    assertTrue(secretCryptoKey.isRevoked());
  }

  @Test
  void testClose() {
    secretCryptoKey.close();
    verify(secretMock).close();
  }

  @Test
  void testGetSecret() {
    Secret actualSecret = secretCryptoKey.getSecret();
    assertEquals(secretMock, actualSecret);
  }

}
